Set top-level window class name

gbrsales opened the pull request:
I'm developing an application for both macOS and Linux (specifically, Ubuntu 18.04). I noticed that the app name is not displayed properly in the task bar or in system menus: only its first character appears there.

In the case of macOS I managed to fix that by calling `SetAppName()` on startup. The Linux story was not so simple. It looks like gnome-shell relies, among other things, on the value of the WM_CLASS property of the top-level window. Indeed, an older version of the GTK-specific code used to set WM_CLASS.
https://github.com/wxWidgets/wxWidgets/blob/fb5c13ed00490bb61a7e77b7a21687381cd70efb/src/gtk1/toplevel.cpp#L538-L539

The function `gtk_window_set_wmclass()` was eventually deprecated. Still, no direct replacement seems to be available. I therefore suggest to reintroduce the call to allow the application to control its own name in the task bar. To make development easier, the actual string is taken from the `ClassName` attribute set on the global application object.

I think this is a low-risk update. Unless the user explicitly sets the application `ClassName`  wxWidgets behavior is not changing.

gbrsales committed:
Set top-level window class name

gnome-shell relies on the WM_CLASS property to set the application
name displayed in the task bar. If that is not defined, a possible
incomplete label is shown to the user (in my tests, just the
first letter of the actual name).

Here I reintroduce a call to the `gtk_window_set_wmclass()` function to
explicitly set the class name. That function is now marked as deprecated,
but no replacement seems to be available (see the discussion at [1]).

To make application development easier, the actual name is taken from
the `ClassName` attribute specified on the global application object.

[1] https://stackoverflow.com/questions/9324163/how-to-set-application-title-in-gnome-shell/9343106#9343106

vadz commented:
Thanks, I agree we should do this, but we probably should only do it when using X (see the existing checks using `GDK_WINDOWING_X11` and `GDK_IS_X11_SCREEN()` in the same file) and we also probably need to disable the deprecated warnings we're going to get for using this function (again, see the existing occurrences of `wxGCC_WARNING_SUPPRESS(deprecated-declarations)/wxGCC_WARNING_RESTORE()`).

Could you please update the patch to do it? TIA!

vadz commented:
BTW, I forgot to mention, but the fact that only the first character of the app name appears in the taskbar looks like a bug to me. Do you have any idea where does this come from? I'm afraid we might be passing a wide char string directly to some function expecting UTF-8 or something like this, as this would account for the name being truncated at its first char.

paulcor commented:
I'd like to have a better understanding of what is going on before applying a hack like this. There may be a better way to get the desired effect. And it's not necessary to have any X11 checks, `gtk_window_set_wmclass()` just has no effect if not using X11.

gbrsales commented:
> BTW, I forgot to mention, but the fact that only the first character of the app name appears in the taskbar looks like a bug to me. Do you have any idea where does this come from? I'm afraid we might be passing a wide char string directly to some function expecting UTF-8 or something like this, as this would account for the name being truncated at its first char.

You were right. Indeed a `wchar_t*` was incorrectly being passed to a function expecting a `char*`, resulting in the observed truncation. Only this did not happen directly in wxWidgets, but in the wxHaskell bindings that I'm using. I've just fixed the issue with https://github.com/gbrsales/wxHaskell/commit/7c3f774a74dfe09331f7c76201ea2c8d7f689f5b. Sorry for the confusion.

Still, the original issue remains valid. The name shown in the task bar now corresponds to the (capitalized) executable filename, but I would rather use a more human-readable string possibly including whitespace. WM_CLASS looks like the way to get there.

I understand from @paulcor comment that the `GDK_WINDOWING_X11` check might not be necessary. The use of `gtk_window_set_wmclass()` is not generating any warnings, so we may avoid `wxGCC_WARNING_SUPPRESS(deprecated-declarations)` too?

vadz commented:
Nice to hear that wxHaskell is (still) being maintained!

Sorry for not checking what `gtk_window_set_wmclass()` did in non-X11 backends, I assumed it wouldn't make sense to call it there, but if Paul says that it won't do any harm, you can indeed trust him on this. I'm still not sure why aren't you getting the deprecation warning, is your GTK+ (build) version less than 3.22?

Also, this function doesn't exist at all in GTK 4, AFAIK, so it would be still nice to find another solution, even though this is clearly not the biggest blocker for GTK 4 transition...

Speaking about less trivial things, this [SO answer](https://unix.stackexchange.com/a/494170/20551) seems to say that we should actually be passing the actual application executable name (`argv[0]`) as the first string and only use the name set via `SetAppName()` as the second one. But the spec does seem to confirm that using the latter as the class name is the right thing to do, at least in my quick reading.

gbrsales committed:
Program name as the first string to gtk_window_set_wmclass()

gbrsales commented:
I have updated the call to `gtk_window_set_wmclass()` to provide the program name as the first string. To avoid code duplication, I have extracted some logic from `wxAppConsoleBase::GetAppName()` and created a new method `DetectExeName()`.

>  I'm still not sure why aren't you getting the deprecation warning, is your GTK+ (build) version less than 3.22?

Configure says:
```
checking for GTK+ - version >= 3.0.0... yes (version 3.22.30)
```

paulcor commented:
I have an alternate proposal, see #1855

vadz commented:
@paulcor What are we going to do about this one, finally?

paulcor commented:
I have checked a few different distributions and determined that Gnome seems to have started using the class as a display name some time after 3.28. Has no effect with a Wayland session, which is generally the default now, although not for Ubuntu. I was hoping to find a more general way of getting this behavior by looking at the Gnome sources, but gave up after expending considerable time digging around. AFAICT the expected way to control an app's display name is to use a `.desktop` file.

I'm not in favor of using `wxApp::GetClassName()`, it's vaguely documented and has apparently had no purpose in the last ~20 years, so I'm disinclined to give anyone a reason to start using it now. I think `wxApp::GetAppDisplayName()` would be more appropriate.

This is the change I'm considering making:
```diff
diff --git a/src/gtk/toplevel.cpp b/src/gtk/toplevel.cpp
index 4a38a00b86..f8df03b61e 100644
--- a/src/gtk/toplevel.cpp
+++ b/src/gtk/toplevel.cpp
@@ -25,6 +25,7 @@
 
 #ifndef WX_PRECOMP
     #include "wx/frame.h"
+    #include "wx/app.h"
     #include "wx/icon.h"
     #include "wx/log.h"
 #endif
@@ -629,6 +630,11 @@ bool wxTopLevelWindowGTK::Create( wxWindow *parent,
 
     m_title = title;
 
+#ifndef __WXGTK4__
+    // Gnome uses class as display name
+    gdk_set_program_class(wxTheApp->GetAppDisplayName().utf8_str());
+#endif
+
     // NB: m_widget may be !=NULL if it was created by derived class' Create,
     //     e.g. in wxTaskBarIconAreaGTK
     if (m_widget == NULL)
```

vadz commented:
FWIW I agree that `GetAppDisplayName()` is what should be used for any user-visible strings, so I definitely don't mind doing it like this. Thanks (and sorry to hear that you lost much time for nothing...)!

paulcor commented:
Closed via 2c62ac41c179bca46dc151e3749747f076ef1fb3
